File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/camera/SimpleCamera.java
package com.mapbox.services.android.navigation.v5.navigation.camera;

import com.mapbox.api.directions.v5.models.DirectionsRoute;
import com.mapbox.core.constants.Constants;
import com.mapbox.geojson.LineString;
import com.mapbox.geojson.Point;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * The default camera used by {@link com.mapbox.services.android.navigation.v5.navigation.MapboxNavigation}.
 *
 * @since 0.10.0
 */
public class SimpleCamera extends Camera {

  protected static final int DEFAULT_TILT = 50;
  protected static final double DEFAULT_ZOOM = 15d;

  private List<Point> routeCoordinates = new ArrayList<>();
  private DirectionsRoute initialRoute;

  @Override
  public double tilt(RouteInformation routeInformation) {
    return DEFAULT_TILT;
  }

  @Override
  public double zoom(RouteInformation routeInformation) {
    return DEFAULT_ZOOM;
  }

  @Override
  public List<Point> overview(RouteInformation routeInformation) {
    boolean invalidCoordinates = routeCoordinates == null || routeCoordinates.isEmpty();
    if (invalidCoordinates) {
      buildRouteCoordinatesFromRouteData(routeInformation);
    }
    return routeCoordinates;
  }

  private void buildRouteCoordinatesFromRouteData(RouteInformation routeInformation) {
    if (routeInformation.route() != null) {
      setupLineStringAndBearing(routeInformation.route());
    } else if (routeInformation.routeProgress() != null) {
      setupLineStringAndBearing(routeInformation.routeProgress().directionsRoute());
    }
  }

  private void setupLineStringAndBearing(DirectionsRoute route) {
    if (route.equals(initialRoute)) {
      return; //no need to recalculate these values
    }
    initialRoute = route;
    routeCoordinates = generateRouteCoordinates(route);
  }

  private List<Point> generateRouteCoordinates(DirectionsRoute route) {
    if (route == null) {
      return Collections.emptyList();
    }
    LineString lineString = LineString.fromPolyline(route.geometry(), Constants.PRECISION_6);
    return lineString.coordinates();
  }
}

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/camera/SimpleCamera.kt
package com.mapbox.services.android.navigation.v5.navigation.camera

import com.mapbox.api.directions.v5.models.DirectionsRoute
import com.mapbox.core.constants.Constants
import com.mapbox.geojson.LineString
import com.mapbox.geojson.Point
import com.mapbox.services.android.navigation.v5.utils.extensions.ifNonNull
import java.util.ArrayList

/**
 * The default camera used by [com.mapbox.services.android.navigation.v5.navigation.MapboxNavigation].
 */
open class SimpleCamera : Camera() {

    companion object {
        protected const val DEFAULT_TILT = 50
        protected const val DEFAULT_ZOOM = 15.0
    }

    private var routeCoordinates: List<Point> = ArrayList()
    private var initialRoute: DirectionsRoute? = null

    override fun tilt(routeInformation: RouteInformation): Double {
        return DEFAULT_TILT.toDouble()
    }

    override fun zoom(routeInformation: RouteInformation): Double {
        return DEFAULT_ZOOM
    }

    override fun overview(routeInformation: RouteInformation): List<Point> {
        if (routeCoordinates.isEmpty()) {
            buildRouteCoordinatesFromRouteData(routeInformation)
        }
        return routeCoordinates
    }

    private fun buildRouteCoordinatesFromRouteData(routeInformation: RouteInformation) {
        ifNonNull(routeInformation.route) { route ->
            setupLineStringAndBearing(route)
        } ?: ifNonNull(routeInformation.routeProgress?.directionsRoute()) { directionsRoute ->
            setupLineStringAndBearing(directionsRoute)
        }
    }

    private fun setupLineStringAndBearing(route: DirectionsRoute) {
        if (route == initialRoute) {
            return // no need to recalculate these values
        }
        initialRoute = route
        routeCoordinates = generateRouteCoordinates(route)
    }

    private fun generateRouteCoordinates(route: DirectionsRoute?): List<Point> =
        route?.geometry()?.let { geometry ->
            val lineString = LineString.fromPolyline(geometry, Constants.PRECISION_6)
            lineString.coordinates()
        } ?: emptyList()
}
